File: kirun-java/src/main/java/com/fincity/nocode/kirun/engine/HybridRepository.java
package com.fincity.nocode.kirun.engine;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class HybridRepository<T> implements Repository<T> {

	private Repository<T>[] repos;

	@SafeVarargs
	public HybridRepository(Repository<T>... repositories) {

		this.repos = repositories;
	}

	@Override
	public T find(String namespace, String name) {

		for (Repository<T> repo : this.repos) {
			T s = repo.find(namespace, name);
			if (s != null)
				return s;
		}

		return null;
	}

	@Override
	public List<String> filter(String name) {

		Set<String> result = new HashSet<>();

		for (var repo : this.repos) {
			result.addAll(repo.filter(name));
		}

		return List.copyOf(result);
	}
}

File: kirun-js/src/engine/HybridRepository.ts
import { Repository } from './Repository';

export class HybridRepository<T> implements Repository<T> {
    repos: Repository<T>[];

    constructor(...repos: Repository<T>[]) {
        this.repos = repos;
    }

    public async find(namespace: string, name: string): Promise<T | undefined> {
        for (let repo of this.repos) {
            let s = await repo.find(namespace, name);
            if (s) return s;
        }

        return undefined;
    }

    public async filter(name: string): Promise<string[]> {
        let result = new Set<string>();
        for (let repo of this.repos) {
            (await repo.filter(name)).forEach((e) => result.add(e));
        }
        return Array.from(result);
    }
}
